pkg/receive/handler.go

```diff
@@ -324,29 +326,39 @@ func (h *Handler) forward(ctx context.Context, tenant string, r replica, wreq *p
 	}
 	h.mtx.RUnlock()
 
-	return h.parallelizeRequests(ctx, tenant, replicas, wreqs)
+	n, ec := h.parallelizeRequests(ctx, tenant, replicas, wreqs)
+	// Collect any errors from forwarding the time series.
+	var errs terrors.MultiError
+	for ; n > 0; n-- {
+		if err := <-ec; err != nil {
+			errs.Add(err)
+		}
+	}
+
+	return errs.Err()
 }
 
 // parallelizeRequests parallelizes a given set of write requests.
 // The function only returns when all requests have finished
 // or the context is canceled.
-func (h *Handler) parallelizeRequests(ctx context.Context, tenant string, replicas map[string]replica, wreqs map[string]*prompb.WriteRequest) error {
+func (h *Handler) parallelizeRequests(ctx context.Context, tenant string, replicas map[string]replica, wreqs map[string]*prompb.WriteRequest) (int, <-chan error) {
 	ec := make(chan error)
-	defer close(ec)
-	// We don't wan't to use a sync.WaitGroup here because that
-	// introduces an unnecessary second synchronization mechanism,
-	// the first being the error chan. Plus, it saves us a goroutine
-	// as in order to collect errors while doing wg.Wait, we would
-	// need a separate error collection goroutine.
+
+	// We need the sync group as a second synchronization mechanism as we can
+	// only close the producing channel when all requests have terminated.
+	var wg sync.WaitGroup
 	var n int
+
 	for endpoint := range wreqs {
 		n++
 		// If the request is not yet replicated, let's replicate it.
 		// If the replication factor isn't greater than 1, let's
 		// just forward the requests.
 		if !replicas[endpoint].replicated && h.options.ReplicationFactor > 1 {
+			wg.Add(1)
 			go func(endpoint string) {
-				ec <- h.replicate(ctx, tenant, wreqs[endpoint])
+				defer wg.Done()
+				ec <- errors.Wrap(h.replicate(ctx, tenant, wreqs[endpoint]), "could not replicate write request")
```

Comment:
Should pass error only on error?

Reply:
errors.Wrap returns nil if the input err is nil

Reply:
I am totally aware of that but IMO it's extremely confusing and prone to error. Plus adds major overhead on critical path.